src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -142,6 +142,100 @@ ShardManager::AsyncResult< ShardInfo > HSHomeObject::_seal_shard(ShardInfo const
     return req->result();
 }
 
+bool HSHomeObject::on_pre_commit_shard_msg(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                           cintrusive< homestore::repl_req_ctx >& hs_ctx) {
+    if (hs_ctx != nullptr) {
+        auto ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx);
+        return do_shard_message_pre_commit(lsn, *r_cast< ReplicationMessageHeader* >(header.bytes), ctx->hdr_buf_,
+                                           hs_ctx);
+    }
+    return true;
+}
+
+bool HSHomeObject::do_shard_message_pre_commit(int64_t lsn, ReplicationMessageHeader& header, sisl::blob value,
+                                               cintrusive< homestore::repl_req_ctx >& hs_ctx) {
+    repl_result_ctx< ShardManager::Result< ShardInfo > >* ctx{nullptr};
+    if (hs_ctx != nullptr) {
+        ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx).get();
+    }
+
+    if (ctx) {
+        RELEASE_ASSERT(!ctx->promise_.isFulfilled(), "do_shard_message_pre_commit promise is already fulfilled");
+    }
+
+    if (header.corrupted()) {
+        LOGW("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
+        return false;
+    }
+
+    if (crc32_ieee(init_crc32, value.bytes, value.size) != header.payload_crc) {
+        // header & value is inconsistent;
+        LOGW("replication message header is inconsistent with value, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::CRC_MISMATCH)); }
+        return false;
+    }
+
+    auto shard_info = deserialize_shard_info(r_cast< const char* >(value.bytes), value.size);
+    switch (header.msg_type) {
+    case ReplicationMessageType::SEAL_SHARD_MSG: {
+        // we can not release chunk here, since if rollback happens, we can not make sure we can get the same chunk.
+        // chunk selector will always return the a chunk of least used, and GC will happen at the time window
+        // we can not make sure we can get the same chunk as before, so we need to wait for the commit phase to release
+        // chunk;
+        update_shard_in_map(shard_info);
+        if (ctx) { ctx->promise_.setValue(ShardManager::Result< ShardInfo >(shard_info)); }
```

Comment:
we can not set promise here, right? otherwise it will response client that seal shard is successfully, but maybe seal shard message is not replicated to quorum later.

Reply:
I also considered this point earlier.
according to the design of nuraft , if we use pre_commit, we need to reply to client in pre_commit. pls refer to https://github.com/eBay/NuRaft/blob/9de4522beccaa1901d49488fcbb01447a9335ea9/docs/async_replication.md?plain=1#L17

it mentioned in nuraft `However, there are some loosened cases where we want to achieve low latency by sacrificing consistency and resolving conflicts manually`.

so if we use pre_commit to take the real action, there might be a case that client get a success response, but then the action is rollbacked. that means , if we adopt pre_commit, we need accept this conflict. that is the contract.

but,  move it to `on_commit` is also i want to do earlier. i write the code here to see if there is any discussion  about.

Reply:
let`s move it to on_commit